docs/proposals/component-registration.md

```diff
@@ -0,0 +1,337 @@
+	Type    ComponentStatusesConditionType `json:"type"`
+	Status  ConditionStatus                `json:"status"`
+	Message string                         `json:"message,omitempty"`
+	Error   string                         `json:"error,omitempty"`
+}
+
+// ComponentStatuses (and ComponentStatusesList) holds the cluster validation info.
+// Deprecated: replaced by ComponentStatus
+type ComponentStatuses struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	Conditions []ComponentStatusesCondition `json:"conditions,omitempty"`
+}
+
+// ComponentStatusesList describes a list of component statuses.
+// Deprecated: replaced by ComponentList
+type ComponentStatusesList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+
+	Items []ComponentStatuses `json:"items"`
+}
+```
+
+### Kubectl (CLI)
+
+Add a `kubectl get components` command to list the components and their last known conditions.
+
+### Storage
+
+Store component spec/status in etcd, with an apiserver storage wrapper.
+
+### Status Updates
+
+Add a component-controller that regularly probes registered components and updates their stored last known status.
+
+Add self-registration to the components that uses the new Components API (Create call) when the component starts.
+
+Add self-deregistration to the components that uses the new Components API (Remove call) when the component stops gracefully (e.g. after receiving SIGTERM).
+
+Add self-updating to the components that uses the new Components API (Update call) when the component errors/panics.
+
+Use a TCP probe to check liveness, because it indicates a server is listening on the expected host/path, even if its responses are unhealthy.
+
+Update individual component `/healthz` endpoints to perform as an indicator of readiness, not just liveness. Readiness should include transitively checking the readiness of external dependencies (e.g. etcd or other databases).
+
+
+## History
+
+The `/componentstatuses` endpoint and associated `kubectl get componentstatuses` CLI command were added to facilitate retrieving the status of the cluster.
+
+Limitations of the current implementation:
+
+- Components are hardcoded
+  - Components assumed to be localhost accessible
+    - Not required/enforced by any other method
+    - Incompatible with HA deployments (deploying components on different machines)
+    - Incompatible with non-host docker networking (deploying components in different containers)
+  - Components assumed to use default ports
+    - Incompatible with custom port configurations
+  - Components assumed to have a root `/healthz` endpoint
+    - Incompatible with components behind reverse proxies with path prefixes
+- Etcd is included as a component
+  - Inconsistent with convention that `/healthz` success implies usability (aka, dependencies are also usable, even if not completely healthy)
+  - Indicates a k8s cluster problem even if only 1 of 3 HA Etcd nodes are down
```

Comment:
see above comment about HA logic. I don't see yet how the proposal addresses this